python/sdist/amici/sbml_utils.py

```diff
@@ -0,0 +1,558 @@
+
+    :param variable_id:
+        SBML ID of the quantity for which the rate rule is to be added.
+
+    :param formula:
+        Formula for the rate rule (it will be sympified).
+
+    :param rule_id:
+        SBML ID of the new rate rule.
+        Defaults to `'rate_' + variableId`.
+
+    :return:
+        The new rate rule as a :py:class:`libsbml.RateRule` object.
+    """
+    variable_id = str(variable_id)
+    if rule_id is None:
+        rule_id = 'rate_' + variable_id
+
+    # Check whether rules exists for this parameter or with the same name
+    # TODO the resulting SBML may still be invalid
+    #      if other types of objects (e.g., species) have the same ID
+    if model.getRuleByVariable(variable_id):
+        raise SbmlDuplicateComponentIdError(
+            f'A rule for parameter {variable_id} has already been defined.'
+        )
+    if model.getRule(rule_id):
+        raise SbmlDuplicateComponentIdError(
+            f'A rule with SBML ID {rule_id} has already been defined.'
+        )
+
+    rule = model.createRateRule()
+    if rule.setVariable(variable_id) != libsbml.LIBSBML_OPERATION_SUCCESS:
+        raise SbmlInvalidIdSyntax(f'{variable_id} is not a valid SBML ID')
+    if rule.setIdAttribute(rule_id) != libsbml.LIBSBML_OPERATION_SUCCESS:
+        raise SbmlInvalidIdSyntax(f'{rule_id} is not a valid SBML ID')
+    set_sbml_math(rule, formula)
+
+    return rule
+
+
+def add_inflow(
+    model: libsbml.Model,
+    species_id: SbmlID,
+    rate,
+    *,
+    reaction_id: Optional[str] = None,
+    reversible: bool = False,
+) -> libsbml.Reaction:
+    species_id = str(species_id)
+    if reaction_id is None:
+        reaction_id = f'inflow_of_{species_id}'
+
+    if model.getReaction(reaction_id):
+        raise SbmlDuplicateComponentIdError(
+            f'A reaction with SBML ID {reaction_id} has already been defined.'
+        )
+
+    reaction = model.createReaction()
+    if reaction.setId(reaction_id) != libsbml.LIBSBML_OPERATION_SUCCESS:
+        raise SbmlInvalidIdSyntax(f'{reaction_id} is not a valid SBML ID')
+    reaction.setReversible(reversible)
+
+    spr = reaction.createProduct()
+    spr.setSpecies(species_id)
+
+    kl = reaction.createKineticLaw()
+    compartment_id = model.getSpecies(species_id).getCompartment()
+    set_sbml_math(kl, sp.Symbol(compartment_id) * rate)
+
+    return reaction
+
+
+###############################################################################
+
+
+def get_sbml_units(model: libsbml.Model, x: Union[SbmlID, sp.Basic]) \
+        -> Union[None, str]:
+    """Try to get the units for expression `x`.
+
+    :param model:
+        SBML model.
+    :param x:
+        Expression to get the units of.
+    :return:
+        A string if the units could be determined, otherwise `None`.
+    """
+    # TODO can the SBML unit inference machinery be used?
+    x = sp.sympify(x)
+    if not x.is_Symbol:
+        return None
+    if x.name == 'time':
```

Comment:
amici_time_symbol / sbml_time_symbol

Reply:
Changed to `x.name == sbml_time_symbol.name`. Not sure if equality of symbols is advisable, I am afraid of sympy assumptions complicating things